src/arithmetic/algebraic_expression.c

```diff
@@ -371,84 +380,60 @@ AlgebraicExpression **AlgebraicExpression_From_Query(const AST *ast, Vector *mat
+    assert(operand_count > 1);
+
     AlgebraicExpressionOperand operands[operand_count];
     memcpy(operands, ae->operands, sizeof(AlgebraicExpressionOperand) * operand_count);
 
-    GrB_Matrix A = operands[0].operand;
-    GrB_Matrix C = A;
-
-    if(operand_count == 1) {
-        /* When transposing we must duplicate, as graph matrices
-         * should be considered immutable. */
-        if(operands[0].transpose) {
-            GrB_Matrix transposed;
-            GrB_Index nrows;
-            GrB_Matrix_nrows(&nrows, A);
-            GrB_Matrix_new(&transposed, GrB_BOOL, nrows, nrows);
-            assert(GrB_transpose(transposed, NULL, NULL, A, NULL) == GrB_SUCCESS);
-            C = transposed;
-        }
-    } else {
-        C = NULL;
-        GrB_Descriptor desc;
-        GrB_Descriptor_new(&desc);
-        AlgebraicExpressionOperand leftTerm;
-        AlgebraicExpressionOperand rightTerm;
-
-        /* Multiply right to left,
-         * A*B*C*D,
-         * X = C*D
-         * Y = B*X
-         * Z = A*Y */
-        while(operand_count > 1) {
-            rightTerm = operands[operand_count-1];
-            leftTerm = operands[operand_count-2];
-            
-            if(!C) {
-                GrB_Index nrows;
-                GrB_Index ncols;
-                GrB_Matrix_nrows(&nrows, leftTerm.operand);
-                GrB_Matrix_ncols(&ncols, rightTerm.operand);
-                GrB_Matrix_new(&C, GrB_BOOL, nrows, ncols);
-            }
-
-            // Multiply and reduce.
-            if(leftTerm.transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
-            if(rightTerm.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
-
-            _AlgebraicExpression_Execute_MUL(C, leftTerm.operand, rightTerm.operand, desc);
-
-            // Quick return if C is ZERO, there's no way to make progress.
-            GrB_Index nvals = 0;
-            GrB_Matrix_nvals(&nvals, C);
-            if(nvals == 0) break;
-
-            // Restore descriptor to default.
-            GrB_Descriptor_set(desc, GrB_INP0, GxB_DEFAULT);
-            GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
-
-            // Assign result and update operands count.
-            operands[operand_count-2].operand = C;
-            operands[operand_count-2].transpose = false;
-            operand_count--;
-        }
-
-        GrB_Descriptor_free(&desc);
-    }    
+    GrB_Descriptor desc;
+    GrB_Descriptor_new(&desc);
+    AlgebraicExpressionOperand leftTerm;
+    AlgebraicExpressionOperand rightTerm;
+
+    /* Multiply right to left,
+     * A*B*C*D,
+     * X = C*D
+     * Y = B*X
+     * Z = A*Y */
+    while(operand_count > 1) {
+        rightTerm = operands[operand_count-1];
+        leftTerm = operands[operand_count-2];
+
+        // Multiply and reduce.
+        if(leftTerm.transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+        if(rightTerm.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
```

Comment:
If we're transposing both terms, we can exploit the property:

`C = A'B'` == `C' = BA`
GraphBLAS won't apply this automatically.